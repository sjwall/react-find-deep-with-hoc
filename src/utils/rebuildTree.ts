import {cloneElement, type ReactElement, type ReactNode} from 'react'
import type {FindDeepResult} from './findDeep'

export function rebuildTree(
  result: FindDeepResult,
  children: ReactNode,
): ReactElement {
  let newChildren: ReactNode
  if (result.tree) {
    newChildren = rebuildTree(result.tree, children)
  } else {
    newChildren = children
  }
  const clone = cloneElement(result.base, {
    ...result.base.props,
    children: newChildren,
  })
  return clone
}
