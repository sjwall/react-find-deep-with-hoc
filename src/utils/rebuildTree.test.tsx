import {View, Text} from 'react-native'
import {render} from '@testing-library/react-native'
import type {FindDeepResult} from './findDeep'
import {rebuildTree} from './rebuildTree'

describe('rebuildTree', () => {
  test('builds new tree', () => {
    const a = 'Hello'
    const b = <Text>{a}</Text>
    const c = <View>{b}</View>
    const tree = <View>{c}</View>

    const result: FindDeepResult = {
      found: a,
      foundIndex: 0,
      base: tree,
      parent: b,
      tree: {
        found: a,
        foundIndex: 0,
        base: c,
        parent: b,
        tree: {
          found: a,
          foundIndex: 0,
          base: b,
          parent: b,
        },
      },
    }

    const newTree = rebuildTree(result, 'World')

    expect(render(newTree).toJSON()).toMatchSnapshot()
  })
})
