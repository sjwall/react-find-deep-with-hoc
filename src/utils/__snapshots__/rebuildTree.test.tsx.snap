// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rebuildTree builds new tree 1`] = `
<View>
  <View>
    <Text>
      World
    </Text>
  </View>
</View>
`;
